### PR TITLE
Add `find` function to `Maybe` namespace

### PR DESCRIPTION
**Pull request by dalefrancis88**
#206

### CODE REVIEW

**Comment by evilsoft** [comment]
Initial 👀

File: src/Maybe/find.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/** @license ISC License (c) copyright 2017 original and current authors */
```
Comment (evilsoft) [inline src/Maybe/find.js]: Should be `2018`

File: src/Maybe/find.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const accumulator = fn => (acc, cur) =>
+  !acc.found && predOrFunc(fn, cur) ? { found: true, value: cur } : acc
+
+// find :: Foldable f => ((a -> Boolean) | Pred) -> f a -> Maybe a
+function find(fn, foldable) {
+  if(!isFunction(fn)) {
+    throw new TypeError('find: Function required for first argument')
+  }
+
+  if(isNil(foldable) && !isFoldable(foldable)) {
```
Comment (evilsoft) [inline src/Maybe/find.js]: As a `Foldable` is required...this should throw as the contract is not met.
Also `NaN`, `null` and `undefined` so the `isNil` check is redundant in this case.

File: src/Maybe/find.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+const isFunction = require('../core/isFunction')
+const isFoldable = require('../core/isFoldable')
+
+const accumulator = fn => (acc, cur) =>
+  !acc.found && predOrFunc(fn, cur) ? { found: true, value: cur } : acc
+
+// find :: Foldable f => ((a -> Boolean) | Pred) -> f a -> Maybe a
+function find(fn, foldable) {
+  if(!isFunction(fn)) {
+    throw new TypeError('find: Function required for first argument')
```
Comment (evilsoft) [inline src/Maybe/find.js]: specifically, either a `Pred` or a predicate function is required for the first argument. Should call that out explicitly.

File: src/Maybe/find.spec.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  t.ok(isFunction(find), 'is a function')
+
+  t.ok(isFunction(find(() => true, undefined).either), 'returns a Maybe for undefined value')
+  t.ok(isFunction(find(() => true, []).either), 'returns a Maybe for array value')
+  t.ok(isFunction(find(() => true, fromArray([])).either), 'returns a Maybe for List value')
+
+  t.equal(find(() => true, undefined).either(() => 'Nothing', x => x), 'Nothing', 'returns a Nothing for undefined value')
+  t.equal(find(x => x > 3, [ 1, 2, 3, 4, 5, 6 ]).either(() => 'Nothing', x => x), 4, 'returns the correct value')
+
+  t.equal(find(x => x > 3, [ 0, null, undefined, 4, 5, 6 ]).either(() => 'Nothing', x => x), 4, 'handles bad values in foldable')
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: No need for `either` on these specs.
can get the same result by doing something like:
```js
 t.equal(find(x => x > 3, [ 0, null, undefined, 4, 5, 6 ]).option('Nothing'), 4, 'handles bad values in foldable')
```

File: src/Maybe/find.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  if(isNil(foldable) && !isFoldable(foldable)) {
+    return Maybe.Nothing()
+  }
+
+  const result = foldable.reduce(accumulator(fn), { found: false, value: undefined })
```
Comment (evilsoft) [inline src/Maybe/find.js]: no need to explicitly set `{ value: undefined }`, `{ found: false}` is equivalent.

File: src/Maybe/find.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/** @license ISC License (c) copyright 2017 original and current authors */
+/** @author Dale Francis (dalefrancis88) */
+
+const curry = require('../core/curry')
+const Maybe = require('./index')
```
Comment (evilsoft) [inline src/Maybe/find.js]: This code is compiled, so you we can safely remove the need for name spacing and use desturcutoring like:
```js
const { Nothing, Just } = require('.')
```

File: src/Maybe/find.spec.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  t.throws(fn('string'), err, 'throws if foldable is truthy string')
+  t.throws(fn(false), err, 'throws if arg is false')
+  t.throws(fn(true), err, 'throws if arg is true')
+  t.throws(fn({}), err, 'throws if arg is empty POJO')
+  t.throws(fn({ hi: 'there' }), err, 'throws if arg is non-empty POJO')
+
+  t.end()
+})
+
+test('find works as expected', t => {
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: We should include a section to get some coverage around using a `Pred` datatype.

**Comment by evilsoft** [comment]
Also we will need to expose this function to the main crocks index [here](https://github.com/evilsoft/crocks/blob/master/src/helpers/index.js#L14). Just a heads up also, will be putting up a PR after this is merged to remove the non-core/non-helpers folder functions from the index files.

This will become a problem when we start splitting out the mono-repo. and would require a dependency on `@crocksjs/Maybe` when the user installs `@crocks/helpers`. (in this case)

If the user wanted `Maybe` functions they will have to install `@crocksjs/Maybe`

File: src/Maybe/find.js (hunk 5 above)
Comment (karthikiyengar) [reply src/Maybe/find.js]: I think a loop and an early-exit wouldn't be possible here because this would also need to work on other Foldables, correct?

File: src/Maybe/find.spec.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  t.throws(fn({}), err, 'throws if foldable is empty POJO')
+  t.throws(fn({ hi: 'there' }), err, 'throws if foldable is non-empty POJO')
+
+  t.end()
+})
+
+test('find works as expected', t => {
+  t.ok(isFunction(find), 'is a function')
+
+  t.ok(isFunction(find(() => true, []).either), 'returns a Maybe for array value')
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: All Sum types in `crocks` provide an `either` method so this may not be a good test for.
you can use `isSameType(Maybe, find(() => true, []))`

File: src/Maybe/find.spec.js (hunk 7 above)
Comment (evilsoft) [reply src/Maybe/find.spec.js]: Still need to add specs using the `Pred` datatype.

File: src/Maybe/find.spec.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  t.ok(isFunction(find(() => true, fromArray([])).either), 'returns a Maybe for List value')
+
+  t.equal(find(x => x > 3, [ 1, 2, 3, 4, 5, 6 ]).either(() => 'Nothing', x => x), 4, 'returns the correct value')
+
+  t.equal(find(x => x > 3, [ 0, null, undefined, 4, 5, 6 ]).either(() => 'Nothing', x => x), 4, 'handles bad values in foldable')
+  t.equal(find(x => x > 6, [ 1, 2, 3, 4, 5, 6 ]).either(() => 'Nothing', x => x), 'Nothing', 'returns nothing when value is not found')
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: We need to have some specs around an empty list/array.

File: src/Maybe/find.js (hunk 5 above)
Comment (evilsoft) [reply src/Maybe/find.js]: That is correct @karthikiyengar!

**Comment by evilsoft** [comment]
Still need documentation on this PR as well.

File: docs/src/pages/docs/crocks/Maybe.md [hunk 10]
```diff
@@ -1445,6 +1445,42 @@ Just('so good')
...
 ```
 
+#### find
+
+`crocks/Maybe/find`
+
+```haskell
+find :: Foldable f => ((a -> Boolean) | Pred) -> f a -> Maybe a
+```
+Used to take a foldable and a function or predicate and searchs through it
```
[the remark this answers is not in the document]
Comment (evilsoft) [reply docs/src/pages/docs/crocks/Maybe.md]: Also, we have a link to the `Pred` docs at the bottom, so that refrerence to `Pred` should link to:

<code>[&#96;Pred&#96;][pred]</code>

But only in this document not the README in Maybe

File: src/Maybe/find.spec.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  t.throws(fn({}), err, 'throws if foldable is empty POJO')
+  t.throws(fn({ hi: 'there' }), err, 'throws if foldable is non-empty POJO')
+
+  t.end()
+})
+
+test('find works with predicates', t => {
+  const largeNumber = Pred(isNumber).concat(Pred(x => x > 100))
+
+  t.equal(find(largeNumber, [ 10, '12', 150, 200, 2000 ]).option(0), 150, 'works with predicates')
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: All specs that are defined against the predicate function, should also be duplicated here

File: src/Maybe/find.spec.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  t.throws(fn(true), err, 'throws if foldable is true')
+  t.throws(fn({}), err, 'throws if foldable is empty POJO')
+  t.throws(fn({ hi: 'there' }), err, 'throws if foldable is non-empty POJO')
+
+  t.end()
+})
+
+test('find works with predicates', t => {
+  const largeNumber = Pred(isNumber).concat(Pred(x => x > 100))
+  const nonBoolPred =
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: I need to find a linting rule or write one, or at least put together a style guide for whitespace. But the general guidelines should be:
1. No *more* than 1 line of white space between expression/statement groupings.
1. block content, closed or indented will be surrounded with 1 line of whitespace.
1. single line arrow function assignment should be split (like you have here) between (2) lines at a minimum, with the assignment on the first and the body indented on the following lines.

File: src/Maybe/find.spec.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  t.ok(isSameType(Maybe, findWrapPred(() => true, fromArray([]))), 'returns a Maybe for List value')
+
+  t.equal(findWrapPred(x => x > 3, [ 1, 2, 3, 4, 5, 6 ]).option('Nothing'), 4, 'returns the correct value')
+
+  t.equal(findWrapPred(x => x > 3, [ 0, null, undefined, 4, 5, 6 ]).option('Nothing'), 4, 'handles bad values in foldable')
+  t.equal(findWrapPred(x => x > 6, [ 1, 2, 3, 4, 5, 6 ]).option('Nothing'), 'Nothing', 'returns nothing when value is not found')
+  t.equal(findWrapPred(() => true, []).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items ([])')
+  t.equal(findWrapPred(() => true, new Array()).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (Array)')
+  t.equal(findWrapPred(() => true, fromArray([])).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (List)')
+
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: extra new line here.

File: src/Maybe/find.spec.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+test('find works with predicates', t => {
+  const largeNumber = Pred(isNumber).concat(Pred(x => x > 100))
+  const nonBoolPred =
+    x => Pred(constant(x))
+  const findWrapPred =
+    (fn, foldable) => find(Pred(fn), foldable)
+
+  t.equal(find(largeNumber, [ 10, '12', 150, 200, 2000 ]).option(0), 150, 'works with predicates')
+
+  t.equals(find(nonBoolPred(undefined), []).option('Nothing'), 'Nothing', 'returns false when wrapped function returns undefined')
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: This is an interesting test...Do not know if we need to test this, as we would get the same results if we returned `true`, or any value on an empty Array/List. This is more a spec for an empty list.

And because of the truthy falsey bits, you could write a whole grouping of tests on JS values like

```js
t.equals(find(nonBoolPred(0), [ 3, 4 ]).option('Nothing'), 'Nothing', 'returns Nothing with falsey number')
t.equals(find(nonBoolPred(1), [ 3, 4 ]).option('Nothing'), 3, 'returns first value in a Just with truthy number')
```

But really do not think we need this test, or all of the ones I described above, as the specs below cover the context of this function, and the specs for `predOrFunc` cover the truthy/falsey values...So I think we have enough coverage to rule this test out.

Thoughts?

File: src/Maybe/find.spec.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+  t.end()
+})
+
+test('find works with predicates', t => {
+  const largeNumber = Pred(isNumber).concat(Pred(x => x > 100))
+  const nonBoolPred =
+    x => Pred(constant(x))
+  const findWrapPred =
+    (fn, foldable) => find(Pred(fn), foldable)
```
Comment (evilsoft) [inline src/Maybe/find.spec.js]: You should probably take advantage of the currying with `find` here by defining this using something like:
```js
const wrapPred = fn =>
  find(Pred(fn))

const gt6 =
  wrapPred(x => x > 6)

const T =
  wrapPred(constant(true))
```

This way we get tests around the curried interface as well, and shorten up some of the spec defs.

File: src/Maybe/find.spec.js (hunk 14 above)
Comment (dalefrancis88) [reply src/Maybe/find.spec.js]: Yeah i think it can be removed

File: src/Maybe/README.md [hunk 16]
```diff
@@ -1417,3 +1417,36 @@ Just('so good')
...
 ```
+
+#### find
+
+`crocks/Maybe/find`
+
+```haskell
+find :: Foldable f => ((a -> Boolean) | Pred) -> f a -> Maybe a
+```
+Using a provided predicate function or a `Pred` datatype, `find` takes a `Foldable` instance and  executes for every value in the `Foldable`, skippy empty indexs. `find` then returns the first value it finds that passes the predicate. If found, `find` returns the value in a `Just`, otherwise a `Nothing` is returned.
```
[the remark this answers is not in the document]
Comment (evilsoft) [reply src/Maybe/README.md]: Oh, then please keep it in. I like bits of personality sneaking in

**Comment by evilsoft** [comment]
image

Super :sweet_potato:...Merging.
